agent/engine/containermetadata/write_metadata.go

```diff
@@ -0,0 +1,267 @@
+	if mdDirectoryPath == "" {
+		err = fmt.Errorf("Failed to create metadata: Invalid file path")
+		return binds, err
+	}
+
+	err = os.MkdirAll(mdDirectoryPath, os.ModePerm)
+	if err != nil {
+		err = fmt.Errorf("Failed to create metadata directory at %s: %s", mdDirectoryPath, err.Error())
+		return binds, err
+	}
+
+	// Create metadata file
+	mdFilePath := filepath.Join(mdDirectoryPath, metadataFile)
+	err = ioutil.WriteFile(mdFilePath, nil, 0644)
+	if err != nil {
+		err = fmt.Errorf("Failed to create metadata file at %s: %s", mdFilePath, err.Error())
+		return binds, err
+	}
+
+	// Acquire the metadata then write it in JSON format to the file
+	metadata := acquireMetadataAtContainerCreate(manager.client, manager.cfg, task)
+	err = metadata.writeToMetadataFile(task, container, manager.cfg.DataDir)
+	if err != nil {
+		err = fmt.Errorf("Failed to write to metadata file %s: %s", mdFilePath, err.Error())
+		return binds, err
+	}
+
+	// Add the directory of this container's metadata to the container's mount binds
+	// We do this at the end so that we only mount the directory if there are no errors
+	// This is the only operating system specific point here, so it would be nice if there
+	// were some elegant way to do this for both windows and linux at the same time
+	instanceBind := fmt.Sprintf("%s/%s:%s/%s", manager.cfg.DataDirOnHost, mdDirectoryPath, mountPoint, container.Name)
+	binds = append(binds, instanceBind)
+	return binds, nil
+}
+
+// UpdateMetadata updates the metadata file after container starts and dynamic
+// metadata is available
+func (manager *metadataManager) UpdateMetadata(dockerID string, task *api.Task, container *api.Container) error {
+	// Check if manager has invalid entries
+	var err error
+	if manager.cfg == nil {
+		err = fmt.Errorf("Failed to update metadata: Invalid inputs")
+		return err
+	}
+
+	// Do not update (non-existent) metadata file for internal containers
+	if container.IsInternal {
+		return nil
+	}
+
+	// Verify metadata file exists before proceeding
+	if !mdFileExist(task, container, manager.cfg.DataDir) {
+		err = fmt.Errorf("Failed to update metadata for container %s of task %s: File does not exist", container, task)
+		return err
+	}
+
+	// Get docker container information through api call
+	dockerContainer, err := manager.client.InspectContainer(dockerID, inspectContainerTimeout)
+	if err != nil {
+		err = fmt.Errorf("Failed to inspect container %s of task %s: %s", container, task, err.Error())
+		return err
+	}
+
+	// Ensure we do not update a stopped, dead, or invalid container
+	if dockerContainer == nil || dockerContainer.State.Dead || !dockerContainer.State.FinishedAt.IsZero() {
+		err = fmt.Errorf("Failed ot update metadata for container %s of task %s: Container stopped or invalid")
+	}
+
+	// Acquire the metadata then write it in JSON format to the file
+	metadata := acquireMetadata(manager.client, dockerContainer, manager.cfg, task)
```

Comment:
If the acquisition fails, we can still safely proceed with metadata creation as we will have an empty metadata struct. We do not want to stop metadata acquisition if we get an error at any single stage because we still want all data still able to be acquired. We have already logged errors that occur during the acquisition instead.